azurelinuxagent/common/protocol/extensions_goal_state.py

```diff
@@ -43,615 +35,156 @@ class ExtensionsGoalState(object):
+        raise NotImplementedError()
+
+    @property
+    def activity_id(self):
+        raise NotImplementedError()
+
+    @property
+    def correlation_id(self):
+        raise NotImplementedError()
+
+    @property
+    def created_on_timestamp(self):
+        raise NotImplementedError()
+
+    @property
+    def status_upload_blob(self):
+        raise NotImplementedError()
+
+    @property
+    def status_upload_blob_type(self):
+        raise NotImplementedError()
+
+    def _set_status_upload_blob_type(self, value):
+        raise NotImplementedError()
+
+    @property
+    def required_features(self):
+        raise NotImplementedError()
+
+    @property
+    def on_hold(self):
+        raise NotImplementedError()
 
     def get_redacted_text(self):
         """
         Returns the raw text (either the ExtensionsConfig or the vmSettings) with any confidential data removed, or an empty string for empty goal states.
         """
         raise NotImplementedError()
 
-    def get_required_features(self):
-        return self._required_features
-
-    def get_status_upload_blob(self):
-        return self._status_upload_blob
-
-    def get_status_upload_blob_type(self):
-        return self._status_upload_blob_type
-
     @staticmethod
     def compare(from_extensions_config, from_vm_settings):
         """
         Compares the two instances given as argument and logs a GoalStateMismatch message if they are different.
 
         NOTE: The order of the two instances is important for the debug info to be logged correctly (ExtensionsConfig first, vmSettings second)
         """
-        def raise_error(attribute):
-            message = "Mismatch in ExtensionsConfig (incarnation {0}) and vmSettings (etag {1}).\nAttribute: {2}\n{3} != {4}".format(
-                from_extensions_config.get_id(), from_vm_settings.get_id(),
-                attribute,
-                attrgetter(attribute)(from_extensions_config), attrgetter(attribute)(from_vm_settings)
-            )
-            raise GoalStateMismatchError(message)
-
-        # TODO: HostGAPlugin 112 does not include the status blob; enable these checks once we can ensure a newer version is available. Also enable parsing
-        #       in the _ExtensionsGoalStateFromVmSettings.compare() method
-        # if from_extensions_config.get_status_upload_blob() != from_vm_settings.get_status_upload_blob():
-        #     raise_error("_status_upload_blob")
-        # if from_extensions_config.get_status_upload_blob_type() != from_vm_settings.get_status_upload_blob_type():
-        #     raise_error("_status_upload_blob_type")
-        if from_extensions_config.get_required_features() != from_vm_settings.get_required_features():
-            raise_error("_required_features")
+        def compare_attribute(attribute):
+            from_extensions_config_value = getattr(from_extensions_config, attribute)
+            from_vm_settings_value = getattr(from_vm_settings, attribute)
```

Comment:
My understanding is that this will throw an `AttributeError`  if the attribute isn't present. Is this something we should handle? I'm thinking of something similar to the `NoneType doesn't have...` errors we got before, which is maybe more inscrutable than a custom message.

Reply:
All exceptions from this code are handled. I don't see the need to handle AttributeError specifically. Also, the attribute names are hardcoded and the unit tests ensure typos are caught when testing.